### PR TITLE
Move PackageRestoreDataSource to Microsoft.VisualStudio.ProjectSystem.Managed

### PR DESCRIPTION
**Pull request by tmeschter**
Most pieces of the dataflow chain that detect changes to NuGet-related items (like `PackageReference`, `PackageVersion`, etc.) and "nominate" a package for restore are already in the Microsoft.VisualStudio.ProjectSystem.Managed project. The last major piece is `PackageRestoreDataSource` which is responsible for detecting if changes have actually occurred in the set of relevant data, detecting cycles in the restore operation, and, most importantly, actually nominating the project.

This PR moves `PackageRestoreDataSource` down to Microsoft.VisualStudio.ProjectSystem.Managed. There are two major parts to this:

1. Creating an abstraction, `INuGetRestoreService`, to handle the nomination logic _and_ tracking pending nominations.
2. Abstracting out the other VS services used by `PackageRestoreDataSource` but incidental to its core functionality (e.g., accessing feature flags, raising notifications to the user, etc.).

Related to [AB#1736319](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1736319).

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/project-system/pull/8850)

### CODE REVIEW

**Comment by tmeschter** [comment]
It may be useful to look at the commits one at a time.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/Snapshots/ProjectPropertiesWrapper.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+using NuGet.SolutionRestoreManager;
+using Microsoft.VisualStudio.ProjectSystem.PackageRestore;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.PackageRestore;
+
+/// <summary>
+///     Wraps a collection of <see cref="ProjectProperty"/> instances to implement the <see cref="IVsProjectProperties"/>
+///     interface for NuGet.
+/// </summary>
+internal class ProjectPropertiesWrapper : ImmutablePropertyCollection<IVsProjectProperty, ProjectProperty>, IVsProjectProperties
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/Snapshots/ProjectPropertiesWrapper.cs:12]: Could these `*Wrapper` types be named `Vs*` without conflicts? I think that'd make their purpose clearer. I don't feel strongly about this though. Just an initial impression.

File: tests/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/PackageRestore/Snapshots/RestoreBuilderTests.cs [hunk 2]
```diff
@@ -58,9 +57,9 @@ public void ToProjectRestoreInfo_WhenNuGetRestoreRuleMissing_ReturnsEmpty()
 
             Assert.Empty(result.MSBuildProjectExtensionsPath);
             Assert.Empty(result.OriginalTargetFrameworks);
-            Assert.Equal(1, result.TargetFrameworks.Count);
+            Assert.Single(result.TargetFrameworks);
 
-            var targetFramework = result.TargetFrameworks.Item(0);
+            var targetFramework = result.TargetFrameworks[0];
```
Comment (drewnoakes) [inline tests/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/PackageRestore/Snapshots/RestoreBuilderTests.cs]: `Single` returns the item, so you can just write:

```c#
var targetFramework = Assert.Single(result.TargetFrameworks);
```

**Comment by drewnoakes** [comment]
There's a bunch of usages of `ImmutableList<>` in here that look like (from what I can see) they could be `ImmutableArray<>` instead.

`ImmutableList<>` is a linked list, where each item gets its own heap-allocated object, and the list too, and values can be spread around the heap due to compaction, making enumeration and other collection operations slower. It only really makes sense if we expect a lot of transformations to occur on the collections and I don't see any such transformations (apologies if I've missed them though).

`ImmutableArray<>` offers near-zero overhead on `Array`, which is about as good as you can get in .NET for a collection.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/INuGetRestoreService.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    /// <param name="cancellationToken">
+    ///     A token indicating when the restore operation should be cancelled (e.g. due to the project being
+    ///     unloaded).
+    /// </param>
+    /// <returns>
+    ///     A <see cref="Task{TResult}"/> indicating when the work associated with the restore request, if any,
+    ///     has completed. The <see cref="Task{TResult}.Result"/> will be <see langword="true"/> if the restore
+    ///     ran and completed successfully, and <see langword="false"/> otherwise.
+    /// </returns>
+    public Task<bool> NominateAsync(ProjectRestoreInfo restoreData, IReadOnlyCollection<PackageRestoreConfiguredInput> inputVersions, CancellationToken cancellationToken);
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/INuGetRestoreService.cs]: Remove `public` from interface member (ditto below).

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/INuGetRestoreService.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    public Task UpdateWithoutNominationAsync(IReadOnlyCollection<PackageRestoreConfiguredInput> inputVersions);
+
+    /// <summary>
+    ///     Indicates that there will be no further calls to <see cref="NominateAsync(ProjectRestoreInfo, IReadOnlyCollection{PackageRestoreConfiguredInput}, CancellationToken)"/>
+    ///     or <see cref="UpdateWithoutNominationAsync(IReadOnlyCollection{PackageRestoreConfiguredInput})"/>.
+    /// </summary>
+    /// <remarks>
+    ///     Called as a hint that any work waiting for a possible future nomination can be cancelled.
+    /// </remarks>
+    void UpdatesComplete();
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/INuGetRestoreService.cs]: This name had me a bit confused. Would `NotifyComplete` (and `NotifyFaulted` below) be clearer?

**Comment by drewnoakes** [comment]
> Note the current implementation runs an async task in a "fire and forget" manner which makes it impossible to reliably unit test certain functionality. I've still added the unit tests (which are timing dependent and so _sometimes_ pass) but marked them as skipped. There are probably better approaches that do not require this "fire and forget" call and that's something we should look at for the future.

You could inject a mock for the fault handler and access the task via that. That shouldn't be too tough. Currently it's buried in `_project.Services.FaultHandler`, but I think you could just import/inject an instance directly as it's in the same scope afaik.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.ProjectRestoreInfoSource.cs [hunk 5]
```diff
@@ -172,52 +218,18 @@ protected virtual bool IsSavedNominationOutOfDate(ConfiguredProject activeConfig
...
-                foreach (var configuredInput in configuredInputs)
-                {
-                    if (loadedProject.ProjectConfiguration.Equals(configuredInput.ProjectConfiguration) &&
-                        loadedProject.ProjectVersion.IsLaterThan(configuredInput.ConfiguredProjectVersion))
-                    {
-                        return true;
-                    }
-                }
-            }
+                await _solutionRestoreService4.RegisterRestoreInfoSourceAsync(this, _projectAsynchronousTasksService.UnloadCancellationToken);
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.ProjectRestoreInfoSource.cs]: Looks like this state machine can be avoided (i.e. remove `async` and `await`).

File: tests/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/PackageRestore/PackageRestoreDataSourceTests.cs [hunk 6]
```diff
@@ -10,6 +10,101 @@
...
+
+            var restoreInfo = ProjectRestoreInfoFactory.Create(msbuildProjectExtensionsPath: @"C:\Alpha\Beta");
+            var configuredInputs = PackageRestoreConfiguredInputFactory.Create(restoreInfo);
+
+            await restoreService.UpdateWithoutNominationAsync(configuredInputs);
+
+            Assert.False(nominateCalled);
+        }
+
+        [Fact(Skip = "Cannot ensure that RegisterRestoreInfoSourceAsync has been called when LoadAsync completes.")]
```
Comment (drewnoakes) [inline tests/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/PackageRestore/PackageRestoreDataSourceTests.cs:52]: See earlier comment on a potential way to make these tests deterministic so we can avoid skipping them.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/Snapshots/ProjectPropertiesWrapper.cs (hunk 1 above)
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/Snapshots/ProjectPropertiesWrapper.cs:12]: I'm not happy with the `*Wrapper` names, either. I've been staring at this code for over a week and wrote these types so the names are somewhat irrelevant to me, but I'd like them to be meaningful/helpful for someone who _hasn't_ spent that much time with them.  So if you think `Vs*` would be a better naming scheme then I'll go with that.

File: tests/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/PackageRestore/Snapshots/RestoreBuilderTests.cs (hunk 2 above)
Comment (tmeschter) [reply tests/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/PackageRestore/Snapshots/RestoreBuilderTests.cs]: TIL 👍🏻

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/INuGetRestoreService.cs (hunk 3 above)
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/INuGetRestoreService.cs]: Now I'm scratching my head over why the compiler didn't complain about that...

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/INuGetRestoreService.cs (hunk 4 above)
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/INuGetRestoreService.cs]: Yes, they would. :-)

I feel like naming is one of those things that gets harder the more familiar you are with the code. 🤣

**Comment by tmeschter** [comment]
> > Note the current implementation runs an async task in a "fire and forget" manner which makes it impossible to reliably unit test certain functionality. I've still added the unit tests (which are timing dependent and so _sometimes_ pass) but marked them as skipped. There are probably better approaches that do not require this "fire and forget" call and that's something we should look at for the future.
> 
> You could inject a mock for the fault handler and access the task via that. That shouldn't be too tough. Currently it's buried in `_project.Services.FaultHandler`, but I think you could just import/inject an instance directly as it's in the same scope afaik.

Oh, yes. That would be straightforward.

A more invasive alternative did occur to me. We need the "fire and forget" only because we need to perform an async operation within the synchronous `OnceInitializedOnceDisposed.Initialize()` method. Note this also means the work it is doing--registering the restore source with NuGet--isn't coordinated with the other work it is doing; that's why the unit tests are currently flaky. That registration could instead be done in the `IProjectDynamicLoadComponent.LoadAsync` method, in which case it doesn't need to be fire and forget. I would need to take responsibility for making sure the registration only happened once. And then at that point it isn't clear that I'm actually getting any benefit from deriving from `OnceInitializedOnceDisposed`.

Aside: I'm starting to view the use of `UnconfiguredProject.Services` as an anti-pattern as it tends to hide what services you actually depend on (as opposed to having them listed out in the importing constructor) and it adds some overhead to unit testing.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/INuGetRestoreService.cs (hunk 4 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/INuGetRestoreService.cs]: 100%. I've had to make peace with a lot of terrible naming choices I've made that it's too late to change. Most recent and egregious is `Workspace` in the language service. It's not a workspace at all, from the perspective of Roslyn. It's a project within a workspace.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PackageRestore/PackageRestoreDataSource.cs [hunk 7]
```diff
@@ -275,12 +240,78 @@ private RestoreData CreateRestoreData(ProjectRestoreInfo restoreInfo, bool succe
                 succeeded: succeeded && lastWriteTime != DateTime.MinValue);
         }
 
+        protected virtual bool IsProjectConfigurationVersionOutOfDate(IReadOnlyCollection<PackageRestoreConfiguredInput> configuredInputs)
```
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PackageRestore/PackageRestoreDataSource.cs:243]: nit: The logic for this method is very similar to the `IsSavedNominationOutOfDate` method I commented on above. Any chance for consolidation would be welcomed.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PackageRestore/PackageRestoreDataSource.cs [hunk 8]
```diff
@@ -275,12 +240,78 @@ private RestoreData CreateRestoreData(ProjectRestoreInfo restoreInfo, bool succe
...
             _enabled = true;
 
             EnsureInitialized();
 
+            if (!_wasSourceBlockContinuationSet)
+            {
+                _wasSourceBlockContinuationSet = true;
+                
+                // Inform the NuGet restore service when there will be no further updates so it can cancel any pending work.
+                _ = SourceBlock.Completion.ContinueWith(t =>
```
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PackageRestore/PackageRestoreDataSource.cs:302]: Q: Is this a pattern now to assign a discard for methods that return a value instead of just calling them like a `void` method?

**Review by MiYanni — APPROVED** [review]

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/INuGetRestoreService.cs (hunk 3 above)
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/INuGetRestoreService.cs]: Ah, right.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/NuGetRestoreService.cs [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+                out IComparable latestSavedVersionForActiveConfiguredProject) ||
+            activeConfiguredProject.ProjectVersion.IsLaterThan(latestSavedVersionForActiveConfiguredProject))
+        {
+            return true;
+        }
+        
+        if (_savedNominatedConfiguredVersion.Count == 1)
+        {
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/NuGetRestoreService.cs:209]: No, this is in the right place. The first `if` statement checks if the current version of the _active_ `ConfiguredProject` is newer than the version that corresponds to what we previously nominated. If so, things are out of date.

If we make it to the second `if` we know the active `ConfiguredProject` isn't out of date. If that's the only configuration we care about then we're done. In a multi-targeting project, however, we need to check the other configurations, which happens in the subsequent `foreach` loop.

Basically the first two `if` statements serve as a fast path through the common case, and we don't have to touch `UnconfiguredProject.LoadedConfiguredProjects`.

[the remark this answers is not in the document]
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/NuGetRestoreService.cs:209]: (Also note this method is unchanged from the original code. So there might be bugs in it, but at least they are the same bugs. :-) )

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PackageRestore/PackageRestoreDataSource.cs (hunk 8 above)
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PackageRestore/PackageRestoreDataSource.cs:302]: It's a language feature, actually. You can assign to "_" without having to declare, and it signals that a value is being returned but you've made a conscious choice to ignore it.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/Snapshots/ProjectPropertiesWrapper.cs (hunk 1 above)
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/Snapshots/ProjectPropertiesWrapper.cs]: Done.

File: tests/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/PackageRestore/Snapshots/RestoreBuilderTests.cs (hunk 2 above)
Comment (tmeschter) [reply tests/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/PackageRestore/Snapshots/RestoreBuilderTests.cs]: Done.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/INuGetRestoreService.cs (hunk 3 above)
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/INuGetRestoreService.cs]: Done.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/INuGetRestoreService.cs (hunk 4 above)
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/INuGetRestoreService.cs]: Done.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.ProjectRestoreInfoSource.cs (hunk 5 above)
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.ProjectRestoreInfoSource.cs]: Don.e

File: tests/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/PackageRestore/PackageRestoreDataSourceTests.cs (hunk 6 above)
Comment (tmeschter) [reply tests/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/PackageRestore/PackageRestoreDataSourceTests.cs]: Done.
